### PR TITLE
Dirty code detection and editor updating issue fixes

### PR DESCRIPTION
**Pull request by thusithak**
### Proposed changes in this pull request

- Editor script now correctly syncs up with UI steps
- Dirty code detection and warning message added (Doesn't warn on template add)

### When should this PR be merged

ASAP

### Follow up actions

None

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [x] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js [hunk 1]
```diff
@@ -645,4 +643,52 @@ function setAttributeStep(element) {
...
     $(element).attr('checked', true);
-}
\ No newline at end of file
+}
+
+var documentBeforeChange;
+
+doc.on("beforeChange", function (document, changeObj) {
+    documentBeforeChange = editorContent;
+    documentBeforeChange = documentBeforeChange.replace(/\n|\t/g, '');
```
Comment (pulasthi7) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js]: check if this replace all the characters

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js [hunk 2]
```diff
@@ -645,4 +643,52 @@ function setAttributeStep(element) {
...
+                };
+
+            // you can trigger a before show if you want
+            obj.trigger('beforeShow');
+
+            // now use the old function to show the element passing the new callback
+            _oldShow.apply(obj, [speed, newCallback]);
+        });
+    }
+});
```
Comment (pulasthi7) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js]: add new line

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js (hunk 1 above)
Comment (thusithak) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js]: fixed with e578878 
Now it removes the new lines, tabs first and removes spaces from the string

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js (hunk 2 above)
Comment (thusithak) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js]: Added with 519182c

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js [hunk 3]
```diff
@@ -76,27 +80,45 @@ function showAllErrors() {
...
+        var stepsInUI = getExecuteStepsInUI();
+        var stepsInScript = getExecuteStepsInScript();
+
+        if (stepsInUI.length < stepsInScript.length){
+            CARBON.showConfirmationDialog('Total number of steps are smaller than that of the Script.',
+                submitFormWithDIsabledScript, null);
+        } else if (stepsInUI.length > stepsInScript.length){
+            CARBON.showConfirmationDialog('Total number of steps are greater than that of the Script.',
+                submitFormWithEnabledScript, null);
+        } else{
```
Comment (pulasthi7) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js]: formatting

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js [hunk 4]
```diff
@@ -76,27 +80,45 @@ function showAllErrors() {
...
-                submitFormWithoutCheck, null);
-        } else {
-            submitForm();
+            CARBON.showConfirmationDialog('Save script with errors? (Script will not be evaluated and will only be' +
+                ' saved)',
+                submitFormWithDIsabledScript, null);
         }
-    });
+        if (warnCount > 0) {
+            CARBON.showConfirmationDialog('Save script with warnings? (Script will not be evaluated and will only be' +
```
Comment (pulasthi7) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js]: will be evaluated

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js [hunk 5]
```diff
@@ -645,4 +658,81 @@ function setAttributeStep(element) {
...
+});
+
+doc.on("change", function (document, changeObj) {
+    var documentAfterChange = document.getValue();
+    documentAfterChange = documentAfterChange.replace(/(?:\r\n|\r|\n)/g, '').replace(/\s/g, '');
+    if (documentAfterChange === documentBeforeChange) {
+        scriptIsDirty = false;
+    } else {
+        scriptIsDirty = true;
+        if (fromTemplateLink || fromStepsAddLink) {
```
Comment (pulasthi7) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/js/configure-authentication-flow.js]: if from template, it's dirty. hence, no need the check

**Comment by pulasthi7** [comment]
@thusithak @Minoli Get the above comments addressed in a separate PR

**Pull request merged**